pkg/functions/runner.go

```diff
@@ -151,44 +147,56 @@ func runGo(ctx context.Context, job *Job) (err error) {
 	return
 }
 
-func waitFor(job *Job, timeout time.Duration) error {
-	url := fmt.Sprintf("http://%s:%s/%s", job.Host, job.Port, readinessEndpoint)
+func waitFor(ctx context.Context, job *Job, timeout time.Duration) (err error) {
+	var (
+		uri      = fmt.Sprintf("http://%s:%s%s", job.Host, job.Port, readinessEndpoint)
+		interval = 500 * time.Millisecond
+		to       = time.After(timeout)
+	)
+
 	if job.verbose {
-		fmt.Printf("Waiting for %v\n", url)
+		fmt.Printf("Waiting for %v\n", uri)
 	}
-	to := time.After(timeout)
+
 	for {
-		select {
-		case <-to:
-			return ErrRunTimeout{timeout}
-		case <-time.After(500 * time.Millisecond):
-			if checkReady(url, job.verbose) {
-				return nil
-			}
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		req, err := http.NewRequestWithContext(ctx, "GET", uri, nil)
+		if err != nil {
+			return fmt.Errorf("error creating readiness check context. %w", err)
 		}
-	}
-}
 
-func checkReady(url string, verbose bool) bool { // checks if the instance has become available
-	resp, err := http.Get(url)
-	if err != nil {
-		if verbose {
-			fmt.Printf("Not ready (%v)\n", err)
+		res, err := http.DefaultClient.Do(req)
+		if err != nil {
+			if err, ok := err.(net.Error); ok && err.Timeout() {
+				return ErrRunTimeout{timeout}
+			}
+			if job.verbose {
+				fmt.Printf("endpoint not available. %v\n", err)
+			}
+			// will continue
+		} else if res.StatusCode != 200 {
+			if job.verbose {
+				fmt.Printf("endpoint returned HTTP %v:\n", res.StatusCode)
+				dump, _ := httputil.DumpResponse(res, true)
+				fmt.Println(string(dump))
+			}
+			// will continue
 		}
-		return false
-	}
-	defer resp.Body.Close()
 
-	if resp.StatusCode == 200 {
-		return true
-	}
+		if res != nil {
+			_ = res.Body.Close()
+		}
 
-	if verbose {
-		fmt.Printf("Endpoint returned HTTP %v.\n", resp.StatusCode)
-		dump, _ := httputil.DumpResponse(resp, true)
-		fmt.Println(dump)
+		select {
+		case <-to:
+			return ErrRunTimeout{timeout}
+		case <-ctx.Done():
+			return ErrContextCanceled
+		case <-time.After(interval):
```

Comment:
This seems to be redundant and racy as the ctx will be cancelled at the same time.